ci-operator/step-registry/idp/freeipa/idp-freeipa-commands.sh

```diff
@@ -0,0 +1,349 @@
+          value: "1"
+        - name: SYSTEMD_NSPAWN_API_VFS_WRITABLE
+          value: "network"
+        ports:
+          - name: http-tcp
+            protocol: TCP
+            containerPort: 80
+          - name: https-tcp
+            protocol: TCP
+            containerPort: 443
+        volumeMounts:
+          - name: freeipa-users
+            mountPath: /users
+        lifecycle:
+          postStart:
+            exec:
+              command:
+              - /bin/sh
+              - -c
+              - |
+                # Redirect all output to stdout/stderr for logging
+                exec > /proc/1/fd/1 2>/proc/1/fd/2
+                # Wait Until FreeIPA service is ready
+                until grep -q "FreeIPA server configured" /data/var/log/ipa-server-configure-first.log; do sleep 30; done
+                # Add User for IPA
+                echo "$IPA_ADMIN_PASSWORD" | kinit admin
+                while IFS="," read -r username password; do echo "$password" | ipa user-add "$username" --first="$username" --last="User" --password-expiration=20320502135053Z --password; done < /users/users.txt
+                # Update certificate for FreeIPA IPA inside openshift-config namespace
+                TOKEN=$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)
+                API_SERVER="https://kubernetes.default.svc"
+                CERT=$(awk '{printf "%s\\n", $0}' /etc/ipa/ca.crt | sed '$ s/\\n$//')
+                if curl -sk -H "Authorization: Bearer $TOKEN" "$API_SERVER/api/v1/namespaces/openshift-config/configmaps/ipa-ca" | grep -q 'NotFound'; then
+                  echo "ConfigMap does not exist, creating it..."
+                  curl -k -X POST "$API_SERVER/api/v1/namespaces/openshift-config/configmaps" \
+                    -H "Authorization: Bearer $TOKEN" \
+                    -H 'Content-Type: application/json' \
+                    -d "{
+                          \"apiVersion\": \"v1\",
+                          \"kind\": \"ConfigMap\",
+                          \"metadata\": {
+                            \"name\": \"ipa-ca\"
+                          },
+                          \"data\": {
+                            \"ca.crt\": \"${CERT}\"
+                          }
+                        }"
+                else
+                  echo "ConfigMap exists, Patching latest certificate"
+                  EXISTING_CERT=$(curl -sk -H "Authorization: Bearer $TOKEN" "$API_SERVER/api/v1/namespaces/openshift-config/configmaps/ipa-ca" | grep '"ca.crt":' | sed 's/.*"ca.crt": "\(.*\)".*/\1/')
+                  COMBINED_CERT="${CERT}\\n${EXISTING_CERT}"
+                  PATCH_PAYLOAD=$(echo "{\"data\":{\"ca.crt\":\"${COMBINED_CERT}\"}}")
+                  # Patch the ConfigMap.
+                  curl -sk -X PATCH -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/merge-patch+json" -d "$PATCH_PAYLOAD" "$API_SERVER/api/v1/namespaces/openshift-config/configmaps/ipa-ca"
+                fi
```

Comment:
@deepakpunia , thanks for exploring out this workaround. When the configmap changes, the openshift-authentication namespace's pods will rotate because IDP configuration references the configmap `"name": "ipa-ca"`. This is side effect.

Thus, to be better, I'd still like use fixed CA. As you know, we use IPA service in daily work, so I believe it should be normal to use a fixed CA, otherwise folks require to often download the new CA if the IPA service is maintained/restarted.

Initial exploration finds:
```
[xxia@2024-02-27 14:12:45 CST my]$ oc rsh freeipa-7b464f7865-545cz
# PRESS Tab key to auto complete all ipa commands
# There are several commands related to CA / cert
sh-5.1# ipa<PRESS Tab key>
ipa                                           ...
...
ipa-ca-install                                ...
ipa-cacert-manage                             ...
ipa-cert-fix                                  ...
ipa-certupdate                                ipa-server-certinstall
...
...                                           ipa-server-install
...
...                                           ipactl
...
```

Then searched Google for `ipa-ca-install` and `ipa-server-certinstall`, found some references which might be useful:
https://access.redhat.com/solutions/6765131 `How do I manually renew Identity Management (IPA) certificates`
https://access.redhat.com/documentation/zh-tw/red_hat_enterprise_linux/6/html/identity_management_guide/install-ca-options#install-ca-external `3.4.2. Installing Using an External CA` mentioning `Run the ipa-server-install script, using the --external-ca option` .

Reply:
As per scrum meeting, creating separate Jira issue `https://issues.redhat.com/browse/OCPQE-19635` , I'll explore as well.

Reply:
@xingxingxia Thanks for help, I will also explore from my side as well.